crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -1121,17 +1106,72 @@ impl actor::HcP2p for HolochainP2pActor {
 
     fn publish(
         &self,
-        _dna_hash: DnaHash,
-        _request_validation_receipt: bool,
-        _countersigning_session: bool,
-        _basis_hash: holo_hash::OpBasis,
+        dna_hash: DnaHash,
+        request_validation_receipt: bool,
+        countersigning_session: bool,
+        basis_hash: holo_hash::OpBasis,
         _source: AgentPubKey,
-        _op_hash_list: Vec<DhtOpHash>,
+        op_hash_list: Vec<DhtOpHash>,
         _timeout_ms: Option<u64>,
-        _reflect_ops: Option<Vec<DhtOp>>,
+        reflect_ops: Option<Vec<DhtOp>>,
     ) -> BoxFut<'_, HolochainP2pResult<()>> {
         Box::pin(async move {
-            tracing::error!("publish is currently a STUB in holochain_p2p--deferring since we'll eventually get stuff on gossip anyways...");
+            use crate::types::event::HcP2pHandler;
+
+            // This single function is/was a weird mix of a bunch of strange
+            // operations.
+            //
+            // - inexplicably `reflect_ops` back to holochain
+            // - actually publish `op_hash_list`
+            // - send a notification to peers that we'd like validation receipts
+            //   - moving this to a separate call
+            // - i have no idea at all what the countersigning bool is for
```

Comment:
Just guessing...maybe the countersigning bool has something to do with not publishing Ops related to a session that's still pending to prevent forks?

Reply:
Not to prevent forks exactly, it's used to send the countersigned actions to witnesses or to the enzyme. They gather signatures and push the set back to the countersigners. The countersigned actions are otherwise withheld until the end of the session. There is no good reason other than it being a lot of plumbing previously for this to be part of publish. It could just as easily be its own operation now.